### PR TITLE
writer: split stats writer payloads based on number of entries.

### PR DESCRIPTION
**Pull request by vlad-mh**
Adds a splitting logic in the stats writer in order never to exceed a pre-configured maximum number of "entries" in a stats payload.

Doing so ensures we don't hit the maximum payload size of the Datadog trace intake API.

The first commit is just a cleanup of the stats writer code, and some added comments.

The second commit adds the configuration value `max_entries_per_payload`. This value should be tweaked very cautiously. Setting it too low will have the agent send much more payloads than it should when flushing, which will impact performances and will hit the backend significantly more. Setting the value too high, and the payloads might end up being rejected due to their size.

The third commit is the actual splitting logic + unit tests.

### CODE REVIEW

**Comment by gbbr** [comment]
@vlad-mh can I review or is it still WIP?

File: writer/stats_writer.go [hunk 1]
```diff
@@ -14,151 +14,294 @@ import (
...
+
+	if maxEntriesPerPayloads <= 0 || nbEntries < maxEntriesPerPayloads { // Nothing to do, break early.
+		return []*model.StatsPayload{&model.StatsPayload{
+			HostName: w.hostName,
+			Env:      w.env,
+			Stats:    stats,
+		}}, len(stats), nbEntries
+	}
+
+	nbPayloads := nbEntries / maxEntriesPerPayloads
```
Comment (LotharSee) [inline writer/stats_writer.go]: As this is integer division, shouldn't it be `nbEntries / maxEntriesPerPayloads + 1`?

**Comment by vlad-mh** [comment]
@gbbr ready for review

File: writer/stats_writer.go (hunk 1 above)
Comment (LotharSee) [reply writer/stats_writer.go]: Oops saw next line ; forget it.

File: writer/stats_writer.go [hunk 2]
```diff
@@ -14,151 +14,294 @@ import (
...
+	//    from these maps. This allows is to have one stat bucket
+	//    per time window.
+	pMaps := make([]map[timeWindow]model.StatsBucket, nbPayloads)
+	for i := 0; i < nbPayloads; i++ {
+		pMaps[i] = make(map[timeWindow]model.StatsBucket, nbPayloads)
+	}
+
+	// 3. Iterate over each entries of each stats. Add the entry
+	//    to one of the payload container mappings, in a round
+	//    robin fashion. In some edge cases, we can end up having
```
Comment (LotharSee) [inline writer/stats_writer.go]: In which case can we have that? I don't see how that could happen as the concentrator is already aggregating?

File: writer/config/stats_writer.go [hunk 3]
```diff
@@ -4,14 +4,23 @@ import "time"
...
-	UpdateInfoPeriod time.Duration
-	SenderConfig     QueuablePayloadSenderConf
+	MaxEntriesPerPayload int
+	UpdateInfoPeriod     time.Duration
+	SenderConfig         QueuablePayloadSenderConf
 }
 
-// DefaultStatsWriterConfig creates a new instance of a StatsWriterConfig using default values.
+// DefaultStatsWriterConfig creates a new instance of a
+// StatsWriterConfig using default values.
```
Comment (gbbr) [inline writer/config/stats_writer.go]: It should be fine to have this comment one one line... Do you have a set limit you want us to enforce? I was thinking 120 should be ok.

File: writer/config/stats_writer.go [hunk 4]
```diff
@@ -4,14 +4,23 @@ import "time"
...
+	// An entry has an average size of 125 bytes in a compressed
+	// payload. The current Datadog intake API limits a compressed
+	// payload to ~3MB (24,000 entries), but let's have the
+	// default ensure we don't have paylods > 1.5 MB (12,000
+	// entries).
+
 	return StatsWriterConfig{
-		UpdateInfoPeriod: 1 * time.Minute,
-		SenderConfig:     DefaultQueuablePayloadSenderConf(),
+		MaxEntriesPerPayload: 12000,
```
Comment (gbbr) [inline writer/config/stats_writer.go]: How would you feel about moving this default value into a constant, and the comment above as documentation for that constant? I think that would go great together and make everything clearer. As the comment stands now, it's a bit unclear at first what it's referring too because it's not referring to the direct next line.

File: writer/stats_writer.go [hunk 5]
```diff
@@ -14,151 +14,294 @@ import (
...
 )
 
-// StatsWriter ingests stats buckets and flushes their aggregation to the API.
+// StatsWriter ingests stats buckets and flushes them to the API.
 type StatsWriter struct {
-	stats    info.StatsWriterInfo
+	BaseWriter
+
+	// InStats is the stream of stat buckets to send out.
+	InStats <-chan []model.StatsBucket
```
Comment (gbbr) [inline writer/stats_writer.go]: Since this is the _only_ input channel, let's call it the usual `in`. It will pair nicely with the same naming of the trace writer's "in" channel, like we have now ([link here](https://github.com/DataDog/datadog-trace-agent/blob/416acd39e81c5d6a1c8b79fe30abee295d4325b0/writer/trace_writer.go#L37)). It needs not be an exported field as it is not used anywhere. A lot of the identifiers in the agent are exported but aren't used in foreign packages.

File: writer/stats_writer.go [hunk 6]
```diff
@@ -14,151 +14,294 @@ import (
...
+	// eye on things.
+	info info.StatsWriterInfo
+
+	// hostName is the resolved host name on which the agent is
+	// running, to be sent as part of a stats payload.
 	hostName string
-	env      string
-	conf     writerconfig.StatsWriterConfig
-	InStats  <-chan []model.StatsBucket
+	// env is environment this agent is configured with, to be
```
Comment (gbbr) [inline writer/stats_writer.go]:
```go
// env specifies the environment that this agent is configured to. It will be
// sent as part of the stats payload.
```
Please add a blank line above this comment, to keep it consistent.

File: writer/stats_writer.go [hunk 7]
```diff
@@ -14,151 +14,294 @@ import (
...
+	BaseWriter
+
+	// InStats is the stream of stat buckets to send out.
+	InStats <-chan []model.StatsBucket
+
+	// info holds some stats about the writer itself, to keep an
+	// eye on things.
+	info info.StatsWriterInfo
+
+	// hostName is the resolved host name on which the agent is
```
Comment (gbbr) [inline writer/stats_writer.go]: s/is/specifies

File: writer/stats_writer.go [hunk 8]
```diff
@@ -14,151 +14,294 @@ import (
...
 
 	return &StatsWriter{
-		hostName: conf.HostName,
-		env:      conf.DefaultEnv,
-		conf:     writerConf,
-		InStats:  InStats,
 		BaseWriter: *NewBaseWriter(conf, "/api/v0.2/stats", func(endpoint Endpoint) PayloadSender {
 			return NewCustomQueuablePayloadSender(endpoint, writerConf.SenderConfig)
 		}),
+
```
Comment (gbbr) [inline writer/stats_writer.go]: I would remove all the blank lines here. No reason to group things, there's only a couple of them. I've noticed in Go it's generally preferred to have little blank lines most of the time, unless comments intervene or if it absolutely makes sense in large snippets.

File: writer/stats_writer.go [hunk 9]
```diff
@@ -14,151 +14,294 @@ import (
...
 func NewStatsWriter(conf *config.AgentConfig, InStats <-chan []model.StatsBucket) *StatsWriter {
 	writerConf := conf.StatsWriterConfig
 	log.Infof("Stats writer initializing with config: %+v", writerConf)
 
 	return &StatsWriter{
-		hostName: conf.HostName,
-		env:      conf.DefaultEnv,
-		conf:     writerConf,
-		InStats:  InStats,
 		BaseWriter: *NewBaseWriter(conf, "/api/v0.2/stats", func(endpoint Endpoint) PayloadSender {
```
Comment (gbbr) [inline writer/stats_writer.go]: This isn't your code but it looks really funky. I wonder if there's any reason to have a copy here. Probably not. Do you think we could remove the `*`? It might be nicer to initialize the value before the struct declaration and then just assign it.

File: writer/stats_writer.go [hunk 10]
```diff
@@ -14,151 +14,294 @@ import (
...
 func (w *StatsWriter) Start() {
 	w.BaseWriter.Start()
+
+	// main goroutine
 	go func() {
 		defer watchdog.LogOnPanic()
 		w.Run()
 	}()
+
+	// monitoring goroutine
```
Comment (gbbr) [inline writer/stats_writer.go]: I don't think any of these comments add value. The code is simple enough as-is. Can we remove them?

File: writer/stats_writer.go [hunk 11]
```diff
@@ -14,151 +14,294 @@ import (
...
+
+		hostName: conf.HostName,
+		env:      conf.DefaultEnv,
+
+		conf: writerConf,
 	}
 }
 
-// Start starts the writer.
+// Start starts the different goroutines of the writer, including the
```
Comment (gbbr) [inline writer/stats_writer.go]: Maybe this would suffice:
```go
// Start starts the writer, awaiting stat buckets and flushing them.
```
If anyone is reading the code, they can see that two methods are started on this struct, and those methods have their own documentation. I'd suggest to keep it simple.

File: writer/stats_writer.go [hunk 12]
```diff
@@ -14,151 +14,294 @@ import (
...
-// StatsWriter ingests stats buckets and flushes their aggregation to the API.
+// StatsWriter ingests stats buckets and flushes them to the API.
 type StatsWriter struct {
-	stats    info.StatsWriterInfo
+	BaseWriter
+
+	// InStats is the stream of stat buckets to send out.
+	InStats <-chan []model.StatsBucket
+
+	// info holds some stats about the writer itself, to keep an
```
Comment (gbbr) [inline writer/stats_writer.go]: I'm afraid that "keep an eye on things" isn't quite specific enough to make this field understandable. Maybe something along the lines of:
```go
// info contains various statistics about the writer, which are occasionally sent as metrics to Datadog.
```
Would it be better?

File: writer/stats_writer.go [hunk 13]
```diff
@@ -14,151 +14,294 @@ import (
...
 	hostName string
-	env      string
-	conf     writerconfig.StatsWriterConfig
-	InStats  <-chan []model.StatsBucket
+	// env is environment this agent is configured with, to be
+	// sent as part of the stats payload.
+	env string
 
-	BaseWriter
+	conf writerconfig.StatsWriterConfig
```
Comment (gbbr) [inline writer/stats_writer.go]: I don't feel strongly about this but it might make everything more "symmetrical" to add a comment here too, since you've started this for all the other fields.

File: writer/stats_writer.go [hunk 14]
```diff
@@ -14,151 +14,294 @@ import (
...
+	// monitoring goroutine
+	go func() {
+		defer watchdog.LogOnPanic()
+		w.monitor()
+	}()
 }
 
-// Run runs the main loop of the writer goroutine. If flushes
-// stats buckets once received from the concentrator.
+// Run runs the event loop of the writer' main goroutine. It reads
```
Comment (gbbr) [inline writer/stats_writer.go]: s/writer'/writer's

File: writer/stats_writer.go [hunk 15]
```diff
@@ -14,151 +14,294 @@ import (
...
+// stat buckets from InStats, builds stat payloads and sends them out
+// using the base writer's payloadSender.
```
Comment (gbbr) [inline writer/stats_writer.go]: IMHO it should suffice to say:
```go
// using the base writer.
```

File: writer/stats_writer.go [hunk 16]
```diff
@@ -14,151 +14,294 @@ import (
...
-			go w.updateInfo()
 		case <-w.exit:
 			log.Info("exiting stats writer")
 			return
 		}
 	}
 }
 
-// Stop stops the main Run loop.
+// Stop stops the writer's goroutines.
```
Comment (gbbr) [inline writer/stats_writer.go]:
```go
// Stop stops the writer.
```

File: writer/stats_writer.go [hunk 17]
```diff
@@ -14,151 +14,294 @@ import (
...
 func (w *StatsWriter) Stop() {
 	close(w.exit)
 	w.exitWG.Wait()
+
+	// Closing the base writer, among other things, will close the
```
Comment (gbbr) [inline writer/stats_writer.go]: This comment seems like something that would belong inside the `BaseWriter`'s documentation, not here. Can we please remove it?

File: writer/stats_writer.go [hunk 18]
```diff
@@ -14,151 +14,294 @@ import (
...
 
 	headers := map[string]string{
 		languageHeaderKey:  strings.Join(info.Languages(), "|"),
 		"Content-Type":     "application/json",
 		"Content-Encoding": "gzip",
 	}
 
-	atomic.AddInt64(&w.stats.Bytes, int64(len(data)))
+	for _, p := range payloads {
+		// Synchronously send the paylods one after the other
```
Comment (gbbr) [inline writer/stats_writer.go]: s/paylods/payloads

Please either lowercase the first letter of the comment or add a period at the end, whichever you prefer.

File: writer/stats_writer.go [hunk 19]
```diff
@@ -14,151 +14,294 @@ import (
...
+
+		atomic.AddInt64(&w.info.Bytes, int64(len(data)))
+	}
+}
+
+type timeWindow struct {
+	start, duration int64
+}
+
+// buildPayloads returns a set of payloads to send out, each paylods
```
Comment (gbbr) [inline writer/stats_writer.go]: s/paylods/payload

File: writer/stats_writer.go [hunk 20]
```diff
@@ -14,151 +14,294 @@ import (
...
+// guaranteed to have the number of stats buckets under the given
+// maximum.
+func (w *StatsWriter) buildPayloads(stats []model.StatsBucket, maxEntriesPerPayloads int) ([]*model.StatsPayload, int, int) {
+	if len(stats) == 0 { // It could happen apparently
```
Comment (gbbr) [inline writer/stats_writer.go]: Can we remove this comment? I don't think it adds any value.

File: writer/stats_writer.go [hunk 21]
```diff
@@ -14,151 +14,294 @@ import (
...
+		return []*model.StatsPayload{}, 0, 0
+	}
+
+	// 1. Get an estimate of how many payloads we need, based on
+	//    the total number of map entries (i.e: sum of number of
```
Comment (gbbr) [inline writer/stats_writer.go]: s/i.e/i.e.

File: writer/stats_writer.go [hunk 22]
```diff
@@ -14,151 +14,294 @@ import (
...
+	//    items in the stats bucket's count map). Called N for the
```
Comment (gbbr) [inline writer/stats_writer.go]: Why call it `N` for the rest of the description. It's already called `nbEntries`. It will make more sense to call it that in the other comments too. What do you think?

File: writer/stats_writer.go [hunk 23]
```diff
@@ -14,151 +14,294 @@ import (
...
+	//    the number of entries in the count map is > than the
+	//    number of entries in the distributions maps, so the
+	//    algorithm is correct, but indeed this means we could do
+	//    better.
+	nbEntries := 0
+	for _, s := range stats {
+		nbEntries += len(s.Counts)
+	}
+
+	if maxEntriesPerPayloads <= 0 || nbEntries < maxEntriesPerPayloads { // Nothing to do, break early.
```
Comment (gbbr) [inline writer/stats_writer.go]: Can we please put this comment on the next line?

File: writer/stats_writer.go [hunk 24]
```diff
@@ -14,151 +14,294 @@ import (
...
+	// 2. Create a slice of N maps, mapping a time window (stat +
+	//    duration) to a stat bucket. We will build the payloads
+	//    from these maps. This allows is to have one stat bucket
+	//    per time window.
+	pMaps := make([]map[timeWindow]model.StatsBucket, nbPayloads)
+	for i := 0; i < nbPayloads; i++ {
+		pMaps[i] = make(map[timeWindow]model.StatsBucket, nbPayloads)
+	}
+
+	// 3. Iterate over each entries of each stats. Add the entry
```
Comment (gbbr) [inline writer/stats_writer.go]: Iterate over all entries of each stat...

File: writer/stats_writer.go [hunk 25]
```diff
@@ -14,151 +14,294 @@ import (
...
+	//    to one of the payload container mappings, in a round
+	//    robin fashion. In some edge cases, we can end up having
+	//    a same entry in several inputted stat buckets. We must
```
Comment (gbbr) [inline writer/stats_writer.go]: s/a/the

File: writer/stats_writer.go [hunk 26]
```diff
@@ -14,151 +14,294 @@ import (
...
+	//    check we never overwrite a entry in the new stats
```
Comment (gbbr) [inline writer/stats_writer.go]: check **that** we never over write **an** entry.

File: writer/stats_writer.go [hunk 27]
```diff
@@ -14,151 +14,294 @@ import (
...
+	//    buckets but cleanly merge instead.
+	i := 0
+	for _, sb := range stats {
```
Comment (gbbr) [inline writer/stats_writer.go]: I really love short variable names, but in this case it might make the below code easier to reason about if we were to call it `bucket`, or `buck` if you will. Then we could also have `newbuck` or `newbucket`. Or even `b` if you like that more. It will pair nicer with the doc.

File: writer/stats_writer.go [hunk 28]
```diff
@@ -14,151 +14,294 @@ import (
...
+		tw := timeWindow{sb.Start, sb.Duration}
+
+		for ekey, e := range sb.Counts {
```
Comment (gbbr) [inline writer/stats_writer.go]: Why not the typical `k, v` - or if you want `k, c` (for count)

File: writer/stats_writer.go [hunk 29]
```diff
@@ -14,151 +14,294 @@ import (
...
+	nbPayloads := nbEntries / maxEntriesPerPayloads
+	if nbEntries%maxEntriesPerPayloads != 0 {
+		nbPayloads++
+	}
+
+	// 2. Create a slice of N maps, mapping a time window (stat +
+	//    duration) to a stat bucket. We will build the payloads
+	//    from these maps. This allows is to have one stat bucket
+	//    per time window.
+	pMaps := make([]map[timeWindow]model.StatsBucket, nbPayloads)
```
Comment (gbbr) [inline writer/stats_writer.go]: What's the reasoning behind `pMaps`? Wouldn't `buckets` be simpler and better?

File: writer/stats_writer.go [hunk 30]
```diff
@@ -14,151 +14,294 @@ import (
...
+
+	// 3. Iterate over each entries of each stats. Add the entry
+	//    to one of the payload container mappings, in a round
+	//    robin fashion. In some edge cases, we can end up having
+	//    a same entry in several inputted stat buckets. We must
+	//    check we never overwrite a entry in the new stats
+	//    buckets but cleanly merge instead.
+	i := 0
+	for _, sb := range stats {
+		tw := timeWindow{sb.Start, sb.Duration}
```
Comment (gbbr) [inline writer/stats_writer.go]: s/tw/window

File: writer/stats_writer.go [hunk 31]
```diff
@@ -14,151 +14,294 @@ import (
...
+					newsb.ErrDistributions[ekey] = sb.ErrDistributions[ekey]
+				}
+			}
+			i++
+		}
+	}
 
-	payload := NewPayload(data, headers)
+	// 4. Create the N payloads from the maps.
+	nbStats := 0
```
Comment (gbbr) [inline writer/stats_writer.go]: I'd opt for:
```go
var nbStats, nbEntries int
```

File: writer/stats_writer.go [hunk 32]
```diff
@@ -14,151 +14,294 @@ import (
...
+	nbEntries = 0
+	payloads := make([]*model.StatsPayload, 0, nbPayloads)
+	for _, pm := range pMaps {
```
Comment (gbbr) [inline writer/stats_writer.go]: How about:
```go
for _, b := range buckets {
```

File: writer/stats_writer_test.go [hunk 33]
```diff
@@ -169,6 +170,210 @@ func TestStatsWriter_UpdateInfoHandling(t *testing.T) {
 	assert.Equal(expectedNumErrors, errorsSummary.Sum)
 }
 
+func TestStatsWriter_BuildPayloads(t *testing.T) {
+	assert := assert.New(t)
+
+	// Common case, no duplicate entries
+	{
```
Comment (gbbr) [inline writer/stats_writer_test.go]: Instead of creating these artificial closures, let's remove the comments and use [sub-tests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks).

File: writer/stats_writer_test.go [hunk 34]
```diff
@@ -169,6 +170,210 @@ func TestStatsWriter_UpdateInfoHandling(t *testing.T) {
...
+		assert.Equal(expectedNbPayloads, len(payloads))
+		assert.Equal(expectedNbPayloads, nbStatBuckets)
+		assert.Equal(expectedNbEntries, nbEntries)
+
+		for i := 0; i < expectedNbPayloads; i++ {
+			assert.Equal(1, len(payloads[i].Stats))
+			assert.Equal(expectedNbEntriesByPayload[i], len(payloads[i].Stats[0].Counts))
+		}
+
+		// Assert the counts match
```
Comment (gbbr) [inline writer/stats_writer_test.go]: Let's not leave commented out code.

File: writer/stats_writer_test.go [hunk 35]
```diff
@@ -169,6 +170,210 @@ func TestStatsWriter_UpdateInfoHandling(t *testing.T) {
...
+		// 		for ekey, e := range s.Counts {
+		// 			assert.Equal(expectedCounts[ekey], e.Value)
+		// 		}
+		// 	}
+		// }
+		// Assert the counts match
+		actualCounts := make(map[string]float64, expectedNbEntries)
+		for _, p := range payloads {
+			for _, s := range p.Stats {
+				for ekey, e := range s.Counts {
```
Comment (gbbr) [inline writer/stats_writer_test.go]: Please use `k, v`, `k, c` or `key, count`, etc.

File: writer/stats_writer_test.go [hunk 36]
```diff
@@ -169,6 +170,210 @@ func TestStatsWriter_UpdateInfoHandling(t *testing.T) {
...
+		// time window (start: 0, duration 1e9).
+		stats := []model.StatsBucket{
+			fixtures.RandomStatsBucket(5),
+			fixtures.RandomStatsBucket(5),
+			fixtures.RandomStatsBucket(5),
+		}
+
+		// Remove duplicates so that we have a predictable
+		// state. In another case we'll test with duplicates.
+		expectedNbEntries := 0
```
Comment (gbbr) [inline writer/stats_writer_test.go]: This is duplicated in two tests. IMHO we should try and remove duplication. This test is too long and complex. It should be short and easy to reason about. If possible, let's create some helper methods.

If we are using a fixture method and then iterating over it to change it, that means that we're not using the right method. Let's instead create helper methods for this. Maybe you can add something new to `fixtures` or just keep it in here.

File: writer/stats_writer.go (hunk 1 above)
Comment (ufoot) [inline writer/stats_writer.go]: I tend to prefer `nbPayloads := ((nbEntries - 1) / maxEntriesPerPayloads) + 1` which avoid a test, but your version might be more readable. Just a side note.

File: writer/config/stats_writer.go (hunk 3 above)
Comment (vlad-mh) [reply writer/config/stats_writer.go]: Yeah my editor is set to 80.. I'll change to 120

File: writer/config/stats_writer.go (hunk 4 above)
Comment (vlad-mh) [reply writer/config/stats_writer.go]: makes sense, will do

File: writer/stats_writer.go (hunk 5 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: Had the same train of thought but didn't want to extend the scope of the PR to other files, as this is a publicly exported field and the PR is already big. If you insist, I can?

File: writer/stats_writer.go (hunk 12 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: works for me

File: writer/stats_writer.go (hunk 7 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: 👍

File: writer/stats_writer.go (hunk 6 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: 👍

File: writer/stats_writer.go (hunk 13 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: figured it was self explanatory

File: writer/stats_writer.go (hunk 9 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: will do

File: writer/stats_writer.go (hunk 8 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: as you wish

File: writer/stats_writer.go (hunk 11 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: ok

File: writer/stats_writer.go (hunk 10 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: 👍

File: writer/stats_writer.go [hunk 37]
```diff
@@ -14,151 +14,294 @@ import (
...
 	w.BaseWriter.Stop()
 }
 
 func (w *StatsWriter) handleStats(stats []model.StatsBucket) {
-	numStats := len(stats)
-
-	// If no stats, we can't construct anything
-	if numStats == 0 {
+	payloads, nbStatBuckets, nbEntries := w.buildPayloads(stats, w.conf.MaxEntriesPerPayload)
+	if len(payloads) == 0 {
```
[the remark this answers is not in the document]
Comment (vlad-mh) [reply writer/stats_writer.go]: we test for len of stats in buildPayloads right away, so same result.

File: writer/stats_writer.go (hunk 18 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: ok

File: writer/stats_writer.go (hunk 22 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: yup

File: writer/stats_writer.go (hunk 29 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: pMaps => payloadMapping
each payload is mapped to a new bucket.
agreed something nicer could be found, but i don't like buckets, and payloads only make sense later in the func.

File: writer/stats_writer.go (hunk 2 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: Indeed, it seemed like this couldn't happen with out current implementation. But it did show up right away in unit tests due to how the fixtures are done, and I don't think we should rely too much on the concentrator's logic. I figured a bit of pain now would go a long way in terms of reusability and potential sneaky bug.

File: writer/stats_writer_test.go (hunk 33 above)
Comment (vlad-mh) [reply writer/stats_writer_test.go]: was looking for them and didn't find in the doc, as I was looking for "case" and not sub-test :D perfect, thanks!

File: writer/stats_writer_test.go (hunk 36 above)
Comment (vlad-mh) [reply writer/stats_writer_test.go]: agreed

File: writer/stats_writer_test.go (hunk 34 above)
Comment (vlad-mh) [reply writer/stats_writer_test.go]: oups

File: writer/stats_writer.go (hunk 29 above)
Comment (gbbr) [reply writer/stats_writer.go]: I see. You are mapping buckets to time windows. What about `distribution`? Would that be any better?

File: writer/stats_writer.go (hunk 17 above)
Comment (gbbr) [reply writer/stats_writer.go]: Ok, I see how this works. `Monitor` is not the best method name for an exit channel 😄but that's another story. Let's keep it. Thanks for explaining.

File: writer/stats_writer.go (hunk 13 above)
Comment (gbbr) [reply writer/stats_writer.go]: Oh, it totally is. I just meant that all the other fields have comments and it will look better if this has one too. But it's a purely esthetical thing. You can leave it as it is, I absolutely don't mind at all.

File: writer/stats_writer.go (hunk 5 above)
Comment (gbbr) [reply writer/stats_writer.go]: Ah, my bad. I thought you added this field. You can leave it as is. If you were to change it, it would be super awesome and nice IMO but up to you. You don't have to!

File: writer/stats_writer.go (hunk 31 above)
Comment (vlad-mh) [reply writer/stats_writer.go]: would've loved to but nbEntries is not a redeclaration, just a value reset.

**Comment by vlad-mh** [comment]
Hopefully this version of the PR fits a bit more.

To be clear on the goal: this PR doesn't aim to fix the overall architecture of how we create and flush stats bucket. There might be better ways to architecture it as a whole, and this requires more thought. This PR does, however, try to ensure we don't go above a certain payload size in worse case situations (very high volume services, or high resource cardinality), as the buckets end up being refused by the API, and as a result we have wrong generated metrics.

@ufoot and I chatted IRL about this. My understanding is that the collector relies on time window and having a full view of data during that time window, so we can't trigger a flush earlier than the ticker, based on number of entries. I think overall this logic still holds?

@gbbr I refactored the common code in the tests and used the sub test feature. I went through most of the nits as well.

**Comment by vlad-mh** [comment]
Hopefully the last round of logic changes.
- the default value for the max number of entries is set to 0, thus disabling the feature. This will allow us to opt-in when payloads are being rejected by the API and see if it works well.
- the last commit adds the "splits" field to the writer's info, and report the number of split the agent is doing as a DD metric.

File: writer/stats_writer.go [hunk 38]
```diff
@@ -14,151 +14,291 @@ import (
...
 
-	w.statsClient.Count("datadog.trace_agent.stats_writer.payloads", int64(swInfo.Payloads), nil, 1)
-	w.statsClient.Count("datadog.trace_agent.stats_writer.stats_buckets", int64(swInfo.StatsBuckets), nil, 1)
-	w.statsClient.Count("datadog.trace_agent.stats_writer.bytes", int64(swInfo.Bytes), nil, 1)
-	w.statsClient.Count("datadog.trace_agent.stats_writer.retries", int64(swInfo.Retries), nil, 1)
-	w.statsClient.Count("datadog.trace_agent.stats_writer.errors", int64(swInfo.Errors), nil, 1)
+	// 1. Get an estimate of how many payloads we need, based on the total
+	//    number of map entries (i.e.: sum of number of items in the stats
+	//    bucket's count map).
+	//    NOTE: we use the number of items in the count map as the
```
Comment (LotharSee) [inline writer/stats_writer.go]: As curiosity, how is that possible to have a different count of Counts and Distribution?

Comment (vlad-mh) [reply writer/stats_writer.go]: we will always have more counts than distributions: https://github.com/DataDog/datadog-trace-agent/blob/master/model/statsraw.go#L89
